src/promptflow-evals/promptflow/evals/evaluate/_code_client/code_client.py

```diff
@@ -0,0 +1,75 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+import json
+import logging
+from concurrent.futures.thread import ThreadPoolExecutor
+
+import pandas as pd
+
+from promptflow.evals.evaluate._utils import load_jsonl
+
+LOGGER = logging.getLogger(__name__)
+
+
+class CodeRun:
+    def __init__(self, run, column_mapping, input_data, evaluator_name=None, **kwargs):
+        self.run = run
+        self.column_mapping = column_mapping
+        self.evaluator_name = evaluator_name if evaluator_name is not None else ""
+        self.input_data = input_data
+
+    def get_result_df(self, exclude_inputs=False):
+        result_df = self.run.result(timeout=60 * 60)
+        if exclude_inputs:
+            result_df = result_df.drop(columns=[col for col in result_df.columns if col.startswith("inputs.")])
+        return result_df
+
+
+class CodeClient:
+    def __init__(self):
+        self._thread_pool = ThreadPoolExecutor(thread_name_prefix="evaluators_thread")
+
+    def _calculate_metric(self, evaluator, input_df, column_mapping, evaluator_name):
+        row_metric_futures = []
+        row_metric_results = []
+        input_df = input_df.rename(columns={v: k for k, v in column_mapping.items()}) if column_mapping else input_df
+        for idx, value in enumerate(input_df.to_dict("records")):
+            row_metric_futures.append(self._thread_pool.submit(evaluator, **value))
+
+        for row_number, row_metric_future in enumerate(row_metric_futures):
+            try:
+                row_metric_results.append(row_metric_future.result())
+            except Exception as ex:  # pylint: disable=broad-except
+                msg_1 = f"Error calculating value for a row for metric {evaluator_name}, "
+                msg_2 = f"failed with error {str(ex)} : Stack trace : {str(ex.__traceback__)}"
+                LOGGER.info(msg_1 + msg_2)
+                # If a row fails to calculate, add an empty dict to maintain the row index
+                # This is to ensure the output dataframe has the same number of rows as the input dataframe
+                # pd concat will fill NaN for missing values
+                row_metric_results.append({})
+
+        return pd.concat(
+            [
+                input_df.add_prefix("inputs."),
+                pd.DataFrame(row_metric_results).add_prefix(
+                    f"outputs.{evaluator_name}." if evaluator_name else "outputs."
+                ),
+            ],
+            axis=1,
```

Comment:
Please add ignore_index=True o avoid getting duplicated indices

Reply:
I tried this but since I am already using axis this is causing issues. My guess is it is due to this, got it from the pandas docs

`This is useful if you are concatenating objects where the concatenation axis does not have meaningful indexing information`